### PR TITLE
Fixes #3575 - Get NetworkInfo from Mesos task updates

### LINKED ISSUE #3575: Container IP not correctly display in 0.15.0 and above
I'm using a Docker network plugin that assigns a routable IP address per container on a Docker container.

In Marathon 0.14.2 release the IP address of the container was correctly displayed

```
Host 172.24.197.102
IP Addresses 192.168.0.0
Ports [31830]
Endpoints 172.24.197.102:31830
Service Discovery n/a
Status Started
```

From Marathon 0.15.0 and above, the IP address of the container is no longer displayed:

```
Host 172.24.197.103
Ports [31781]
Endpoints 172.24.197.103:31781
Service Discovery n/a
Status Started
```

Why is the actual container IP no longer available in Marathon?

I'm running Mesos 0.28 on CentOS 7.

### PR DESCRIPTION
**Pull request by gkleiman**
**NOTE**: IP-per-Task is also broken in Marathon v0.15.x. We might need to backport this commit.

- Recent versions of Mesos send the NetworkInfo messages for all tasks,
  even if no ip address was requested.
- The NetworkInfo messages are wrapped inside of the ContainerStatus message.
- Marathon should always store the NetworkInfo messages and expose the
  IpAddresses through the tasks API. Clients (including the UI) can then decide
  if they want to use the agent info or one of the ip addresses.

Changes in order to always return the IpAddreses:

- Refactored the Task.Networking traits. It is now a case class that stores both
  host ports and NetworkInfo messages.
- Changed Task.update in order to always extract the NetworkInfo messages from
  the update messages and store them using new Task.Networking case class.

### CODE REVIEW

**Comment by gkleiman** [comment]
I tested this using the `net-modules` repo and also added a checks to an integration test, to make sure that we always return ip addresses.

File: src/main/scala/mesosphere/marathon/event/Events.scala [hunk 1]
```diff
@@ -223,7 +223,7 @@ case class MesosStatusUpdateEvent(
   message: String,
   appId: PathId,
   host: String,
-  ipAddresses: Seq[org.apache.mesos.Protos.NetworkInfo.IPAddress],
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/Events.scala]: Isn't the order relevant?

File: src/test/scala/mesosphere/marathon/MarathonTestHelper.scala [hunk 2]
```diff
@@ -257,7 +257,7 @@ object MarathonTestHelper {
         status = Task.Status(
           stagedAt = now
         ),
-        networking = Task.HostPorts(Seq(1, 2, 3))
+        networking = Task.Networking(Seq(1, 2, 3))
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/MarathonTestHelper.scala]: I'd prefer named arguments here. (ports = )

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 3]
```diff
@@ -86,11 +86,13 @@ object Task {
...
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, taskStatus: MarathonTaskStatus.Running, now) if !hasStartedRunning =>
         val updated = copy(
+          networking = networking.copy(networkInfos = taskStatus.networkInfos),
           status = status.copy(
             startedAt = Some(now),
-            mesosStatus = mesosStatus))
+            mesosStatus = taskStatus.mesosStatus)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: aren't we essentially storing the network infos twice in our task? Once as part of the status and once as part of the networking? same potentially in the protobuf?

Or am I misunderstanding this?

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 4]
```diff
@@ -101,7 +103,10 @@ object Task {
       case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
         updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus) match {
           case Some(newStatus) =>
-            val updatedTask = copy(status = status.copy(mesosStatus = Some(newStatus)))
+            val updatedTask = copy(
+              networking = networking.copy(networkInfos = taskStatus.networkInfos),
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: Just to be sure, 
* it is not possible that these networkInfos change without a Mesos State change, correct?
* all networkInfos are sent with every statusUpdate? Not only if these change?

File: src/main/scala/mesosphere/marathon/core/task/bus/MarathonTaskStatus.scala [hunk 5]
```diff
@@ -1,13 +1,18 @@
...
+import org.apache.mesos.Protos.{ NetworkInfo, TaskStatus }
 
 sealed trait MarathonTaskStatus {
   def terminal: Boolean = false
-
   def mesosStatus: Option[TaskStatus]
-  def mesosHealth: Option[Boolean] = mesosStatus.flatMap { status =>
-    if (status.hasHealthy) Some(status.getHealthy) else None
+
+  def networkInfos: Seq[NetworkInfo] = {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/bus/MarathonTaskStatus.scala]: In principal, this could also be part of `Task.Status`

File: src/test/scala/mesosphere/marathon/MarathonTestHelper.scala [hunk 6]
```diff
@@ -110,7 +110,7 @@ object MarathonTestHelper {
       .addResources(memResource)
       .addResources(diskResource)
 
-    portsResource.foreach(offerBuilder.addResources(_))
+    portsResource.foreach(offerBuilder.addResources)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/MarathonTestHelper.scala]: why? It's not related to this patch and I don't think that we have a team agreement what we like better.

**Comment by kolloch** [comment]
Maybe we should discuss whether to really duplicate the network info information in our task state. All other issues are just minor.

**Comment by gkleiman** [comment]
@kolloch thanks for the review!

Updated the PR to address your feedback.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 7]
```diff
@@ -115,15 +115,15 @@ trait Formats
     val base = Json.obj(
       "id" -> task.taskId,
       "slaveId" -> task.agentInfo.agentId,
-      "host" -> task.agentInfo.host
+      "host" -> task.agentInfo.host,
+      "ipAddresses" -> task.ipAddresses
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: So it is important to you to have an empty list there disregarding the task state instead of not having the ipAddresses field in case the task is not "launched"?

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 8]
```diff
@@ -73,24 +75,25 @@ object Task {
...
-    override def launched: Option[Launched] = Some(Task.Launched(appVersion, status, networking))
+    override def launched: Option[Launched] = Some(Task.Launched(appVersion, status, hostPorts))
 
     private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
 
     //scalastyle:off cyclomatic.complexity method.length
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, taskStatus: MarathonTaskStatus.Running, now) if !hasStartedRunning =>
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: unnecessary change

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 9]
```diff
@@ -101,7 +104,9 @@ object Task {
       case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
         updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus) match {
           case Some(newStatus) =>
-            val updatedTask = copy(status = status.copy(mesosStatus = Some(newStatus)))
+            val updatedTask = copy(
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: unnecessary change

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 10]
```diff
@@ -181,25 +186,27 @@ object Task {
...
-    override def launched: Option[Launched] = Some(Task.Launched(appVersion, status, networking))
+    override def launched: Option[Launched] = Some(Task.Launched(appVersion, status, hostPorts))
 
     private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
 
     //scalastyle:off cyclomatic.complexity method.length
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, taskStatus: MarathonTaskStatus.Running, now) if !hasStartedRunning =>
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: unnecessary change

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 11]
```diff
@@ -215,7 +222,9 @@ object Task {
       // case 3: health or state updated
       case TaskStateOp.MesosUpdate(_, taskStatus, _) =>
         updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus).map { newStatus =>
-          val updatedTask = copy(status = status.copy(mesosStatus = Some(newStatus)))
+          val updatedTask = copy(
+            status = status.copy(mesosStatus = Some(newStatus))
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: variation: no change necessary ;)

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskSerializer.scala [hunk 12]
```diff
@@ -122,18 +114,12 @@ object TaskSerializer {
     def setReservation(reservation: Task.Reservation): Unit = {
       builder.setReservation(ReservationSerializer.toProto(reservation))
     }
-    def setLaunched(appVersion: Timestamp, status: Task.Status, networking: Task.Networking): Unit = {
+    def setLaunched(appVersion: Timestamp, status: Task.Status, ports: Seq[Int]): Unit = {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskSerializer.scala]: complete nit-pick: hostPorts would be even more consistent

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskSerializer.scala [hunk 13]
```diff
@@ -122,18 +114,12 @@ object TaskSerializer {
...
       builder.setVersion(appVersion.toString)
       builder.setStagedAt(status.stagedAt.toDateTime.getMillis)
       status.startedAt.foreach(startedAt => builder.setStartedAt(startedAt.toDateTime.getMillis))
       status.mesosStatus.foreach(status => builder.setStatus(status))
-      networking match {
-        case Task.HostPorts(hostPorts) =>
-          builder.addAllPorts(hostPorts.view.map(Integer.valueOf(_)).asJava)
-        case Task.NetworkInfoList(networkInfoList) =>
-          builder.addAllNetworks(networkInfoList.asJava)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskSerializer.scala]: I would love a comment in the marathon.proto which explains that the networks info in the protocol buffers is redundant and we got rid of that. Maybe even rename the field to "DEPRECATED_..." or comment it out.

File: src/test/scala/mesosphere/marathon/MarathonTestHelper.scala [hunk 14]
```diff
@@ -361,12 +362,12 @@ object MarathonTestHelper {
...
   def taskLaunched: Task.Launched = {
     val now = Timestamp.now()
-    Task.Launched(now, status = Task.Status(now), networking = Task.NoNetworking)
+    Task.Launched(now, status = Task.Status(now), hostPorts = Seq.empty)
   }
 
   def taskLaunchedOp(taskId: Task.Id): TaskStateOp.LaunchOnReservation = {
     val now = Timestamp.now()
-    TaskStateOp.LaunchOnReservation(taskId, now, Task.Status(now), Task.NoNetworking)
+    TaskStateOp.LaunchOnReservation(taskId, now, Task.Status(now), Seq.empty)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/MarathonTestHelper.scala]: nit-pick, named argument would be nice

File: src/test/scala/mesosphere/marathon/MarathonTestHelper.scala [hunk 15]
```diff
@@ -406,7 +407,7 @@ object MarathonTestHelper {
         startedAt = None,
         mesosStatus = Some(statusForState(taskId, Mesos.TaskState.TASK_STAGING))
       ),
-      Task.NoNetworking
+      Seq.empty
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/MarathonTestHelper.scala]: named argument

File: src/test/scala/mesosphere/marathon/core/appinfo/TaskCountsTest.scala [hunk 16]
```diff
@@ -182,7 +184,7 @@ class Fixture {
       startedAt = None,
       mesosStatus = None
     ),
-    Task.NoNetworking
+    Seq.empty
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/core/appinfo/TaskCountsTest.scala]: as you have realized by now, I really like named arguments in these cases (Seq.empty doesn't give you a hint what the parameter is).

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 17]
```diff
@@ -256,9 +256,12 @@ class AppDeployIntegrationTest
...
 
-    val tasks = marathon.tasks(appId)
-    tasks.code should be(200)
-    tasks.value should have size 2
+    val tasksResult: RestResult[List[ITEnrichedTask]] = marathon.tasks(appId)
+    tasksResult.code should be(200)
+
+    val tasks = tasksResult.value
+    tasks should have size 2
+    tasks.foreach(_.ipAddresses.get should not be empty)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: :+1:

**Comment by kolloch** [comment]
Mostly good to go!

Only major item: Clarify the deliberate non-usage of "networks" in the protocol buffers for future generations of Marathon developers ;)

**Comment by gkleiman** [comment]
@kolloch added an extra commit with the required changes, looking forward to the next round of comments =)

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 18]
```diff
@@ -123,9 +123,10 @@ trait Formats
         "startedAt" -> launched.status.startedAt,
         "stagedAt" -> launched.status.stagedAt,
         "ports" -> launched.hostPorts,
-        "version" -> launched.appVersion,
-        "ipAddresses" -> launched.ipAddresses
-      )
+        "version" -> launched.appVersion
+      ) ++ launched.ipAddresses.map { ipAddresses =>
+          Json.obj("ipAddresses" -> ipAddresses)
+        }.getOrElse(Json.obj())
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: I think that you do not have to make it that complicated. `"ipAddresses" -> launched.ipAddresses` should suffice.

```scala
scala> case class WithOption(test: Option[String])
defined class WithOption

scala> implicit val format = play.api.libs.json.Json.writes[WithOption]
format: play.api.libs.json.OWrites[WithOption] = play.api.libs.json.OWrites$$anon$2@47fc5ae4

scala> play.api.libs.json.Json.toJson(WithOption(None))
res1: play.api.libs.json.JsValue = {}

scala> play.api.libs.json.Json.toJson(WithOption(Some("test")))
res2: play.api.libs.json.JsValue = {"test":"test"}
```

Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: The output of this complicated block and the simpler version are not the same.

The simple version can result in `"ipAddresses": null` if there's no ip address. The more complicated version will skip the property in that case.

I'll update the PR with an intermediate version using foldLeft.

File: src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala [hunk 19]
```diff
@@ -43,35 +43,42 @@ class HealthCheckWorkerActor extends Actor with ActorLogging {
...
-        case HTTP  => http(app, task, launched, check, port)
-        case TCP   => tcp(app, task, launched, check, port)
-        case HTTPS => https(app, task, launched, check, port)
-        case COMMAND =>
-          Future.failed {
-            val message = s"COMMAND health checks can only be performed " +
-              "by the Mesos executor."
-            log.warning(message)
-            new UnsupportedOperationException(message)
+    task.effectiveIpAddress(app) match {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala]: This double nesting is not so nice. I would be nicer to kind of chain the stuff without nesting -- e.g. with `Try`.

**Comment by kolloch** [comment]
Looks great, let's get that shipped. Squash and I'll merge.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 18 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: That sucks but it is not so important. Just ignore my comment then ;)

**Comment by gkleiman** [comment]
Squashed, this includes the following change:

```diff
w/src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala
index bd8b2da..312c6d8 100644
--- c/src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala
+++ w/src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala
@@ -119,14 +119,16 @@ trait Formats
     )

     val launched = task.launched.map { launched =>
-      base ++ Json.obj (
-        "startedAt" -> launched.status.startedAt,
-        "stagedAt" -> launched.status.stagedAt,
-        "ports" -> launched.hostPorts,
-        "version" -> launched.appVersion
-      ) ++ launched.ipAddresses.map { ipAddresses =>
-          Json.obj("ipAddresses" -> ipAddresses)
-        }.getOrElse(Json.obj())
+      launched.ipAddresses.foldLeft(
+        base ++ Json.obj (
+          "startedAt" -> launched.status.startedAt,
+          "stagedAt" -> launched.status.stagedAt,
+          "ports" -> launched.hostPorts,
+          "version" -> launched.appVersion
+        )
+      ){
+        case (launchedJs, ipAddresses) => launchedJs ++ Json.obj("ipAddresses" -> ipAddresses)
+      }
     }.getOrElse(base)

     val reservation = task.reservationWithVolumes.map { reservation =>
```

**Comment by gkleiman** [comment]
@kolloch confirmed it works with Calico's `net-modules`, feel free to hit merge =).

**Comment by kolloch** [comment]
```
       _~
    _~ )_)_~
    )_))_))_)
    _!__!__!_
    \______t/
  ~~~~~~~~~~~~~
```
